### DISCUSSION TITLE
Strategy for multi-tenant deployments?

### DISCUSSION THREAD

**Discussion by slobdellreachout** [gocd/gocd#12691] [General]
Hey guys, I'm new here!  You can thank Octopus Deploy's price restructuring for that.  Spent the day reading docs and playing with my new GOCD instance.  The whole pipeline concept is new to me, but I'm liking it.

My question is about handling multiple tenants.  I'm struggling to find a good way to deal with this with GOCD.  Say I have a website, and I want to deploy it to the same VM 10 times but with different values in the config file, is there a good way to do this?

From what I've seen the only way to do this might be to setup a pipeline for each tenant.  I don't know I'm spit ballin here, help me out.

Steve

**Comment by sriramnrn** [comment]
On Sat, 13 Apr 2024 at 9:30 AM, slobdellreachout ***@***.***>
wrote:

> Hey guys, I'm new here! You can thank Octopus Deploy's price restructuring
> for that. Spent the day reading docs and playing with my new GOCD instance.
> The whole pipeline concept is new to me, but I'm liking it.
>
> My question is about handling multiple tenants. I'm struggling to find a
> good way to deal with this with GOCD. Say I have a website, and I want to
> deploy it to the same VM 10 times but with different values in the config
> file, is there a good way to do this?
>
> From what I've seen the only way to do this might be to setup a pipeline
> for each tenant. I don't know I'm spit ballin here, help me out.
>

You could either:
- set up a pipeline per tenant, with a stage having one job to deploy for
that tenant
- set up a pipeline with a stage having ten parallel jobs each representing
a tenant

When setting up pipelines that differ only in the values passed in, you
should use pipeline templates.

https://docs.gocd.org/current/configuration/pipeline_templates.html

Since you have just begun exploring, have a look at the yaml approach for
configuration as well.

See:
https://docs.gocd.org/current/advanced_usage/pipelines_as_code.html




Steve

**Reply by chadlwilson** [reply]
All worth considering, but pipeline templates do have the downside of not having the template itself be manageable via code/source control, which can feel pretty archaic if you're using to modern build automation systems.

With YAML pipelines as code, you can use yaml anchors and such to reuse snippets across pipelines (bit clunky and with its limitations), or if it's ok for your security posture you could consider something more scriptable like https://github.com/gocd-contrib/gocd-groovy-dsl-config-plugin which allows you to script 'n' pipelines using Groovy, generate pipelines or chain of pipelines from an array of 'tenants' etc.

**Comment by sriramnrn** [comment]
On Sat, 13 Apr 2024 at 10:22 AM, Chad Wilson ***@***.***>
wrote:

> All worth considering, but pipeline templates do have the downside of not
> having the template itself be manageable via code/source control, which can
> feel pretty archaic if you're using to modern build automation systems.
>
My thoughts on the matter:
-  one could push the config XML to version control
- those who are used to designing a template via a graphical UI would still
benefit. There are many installations where once a pipeline workflow is
designed then nothing much changes for years.
- keeping the orchestration separate from build automation is still
required.

> With YAML pipelines as code, you can use yaml anchors and such to reuse
> snippets across pipelines, or if it's ok for your security posture you
> could consider something more scriptable like
> https://github.com/gocd-contrib/gocd-groovy-dsl-config-plugin which
> allows you to script 'n' pipelines using Groovy, generate pipelines or
> chain of pipelines from an array of 'tenants' etc.
>
Definitely worth investigating for those up to it.

**Comment by slobdellreachout** [comment]
Thanks guys, I think I have it worked out in my head!
